src/macaron/repo_finder/provenance_finder.py

```diff
@@ -35,191 +37,350 @@ def __init__(self) -> None:
                 elif isinstance(registry, JFrogMavenRegistry):
                     self.jfrog_registry = registry
 
-    def find_provenance(self, purl: PackageURL) -> InTotoPayload | None:
-        """Find the provenance files of the passed PURL.
+    def find_provenance(
+        self,
+        purl: PackageURL,
+        discovery_functions: list[Callable[..., list[InTotoPayload]]] | None = None,
+        parameters: list[Any] | None = None,
```

Comment:
From the implementation of this method, I can see that:
- `parameters` is provided as positional arguments to `discovery_function`. 
- `parameters` is provided the same way for each function in `discovery_function`.
This mean that the interface of each function must be the same. If their interfaces are different (e.g the order of arguments are different, the number of arguments are different, etc.) we can have a lot issues.
Is is okay to assume that all discovery functions are going to have the same interface and return type signature? :thinking:

Reply:
You raise a good point. I think it would make more sense to have a separate set of parameters for each function call.